### DISCUSSION TITLE
Stabilizing Iterator behaviour

### DISCUSSION THREAD

**Discussion by realpha** [realpha/eitherway#1] [Ideas]
**Motivation:** With the [Iterator Hepers Proposal](https://github.com/tc39/proposal-iterator-helpers) currently at Stage 3, and how popular the pattern has been in other ecosystems, I assume that `Iterable`s will get adopted more widely over the coming years. Taking cues from `Rust`, where both `Result<T, E>` and `Option<T>`  implement the `Iterator` trait and how productive this inter-op can be, I wanted to add support for the iterator protocol from the start. 

**Current Behaviour**: The current implementations all follow the same logic. Every type has an explicit `.iter()` method, which returns an `IterableIterator<T>` over the encapsulated value of type `T` if the `Result`/`Option` is `Ok`/`Some`. The failure variants `(Err` and `None`) return an empty `IterableIterator`.
Additionally, every type inherently supports the iterator protocol via `Symbol.iterator` IF the encapsulated value of type `T` is an `Iterable` by delegating to the implementation of the encapsulted value. Again, the failure variants return an empty `IterableIterator`. 
`Task<T, E>` follows the same logic, but returns `AsyncIterableIterator<T>` or `AsyncIterableIterator<T extends AsyncIterable<infer U> ? U : never>` respectively.

**IMO:** Personally, I am quite fond of how this works. Especially in the case  of `Option<T>`, this implementation strategy yields concise usage patterns, which simplify dealing with with edge cases and have a nice symmetry. An example is the support of the spread operator:
```typescript
const someArr = Some([1, 2, 3, 4]);
const noneArr = None;
const someCopy = [ ...someArr ];
const noneCopy = [ ...noneArr ];

assertEquals(someCopy, [1, 2, 3, 4]);
assertEqulas(noneCopy, []);
```
At the same time, I think (and it seems like I am not alone) that this behaviour can be a little weird when dealing with `string`s because it's the only primitive type which supports the iterator protocol and yields the unicode code points of the `string`, which are typed as `string` respectively. 
Of course this is the standard behaviour of the string type under iteration, nevertheless I think in this particular situation, the dual API may not be unsurprising and detrimental to the API goal of safety.

To sum it up, the current behavior exhibits the following:

**Pros**  ✅:
- Inter-op with built-in operators and collection factories
- Forward compatibility as the language and ecosystem evolves
- Flexibility when dealing `Iterable` types
- Fully type-safe
- Low maintenance effort, as types and implementation are fairly trivial

**Cons** ⛔️:
- The dual API
- Might violate the principle of *least surprise*
- The feature might not be that important to the users of `eitherway`

I would be in favor to leave it as it, but am totally open to suggestions and points I haven't considered.

**Comment by realpha** [comment]
I think we are good here
